server/models/segment.model.js

```diff
@@ -136,6 +138,27 @@ module.exports = function(mongoose) {
           associationName: 'tags',
           payload: tagsToRemove,
         }));
+
+      //Update segmentCount for tags.
+      const tagsToCountSegmentsFor = (
+        await RestHapi.list({
+          model: 'tag',
+          query: {
+            name: currentTags.map((t) => t.tag.name),
```

Comment:
It looks like there will be an issue here since only the tags from the "currentTags" list will be updated. This means any tags that were removed will only be in "oldTags" and therefore won't be updated.

I see two options to fix this:

1) Include the names for both `oldTags` and `currentTags`
2) Instead of updating them all in one loop, update each tag immediately after it is modified. This will also remove the need to query the db for the tags as you will already have them. Basically, for every tag in the `newTagsToAdd` or `existingTagsToAdd` lists, increase their segment count by 1, and do the opposite for the tags in the `deletedTags` list.

Reply:
This is also a great example of how unit/e2e tests can be helpful. I've updated the tests to check for segment count changes when adding and removing tags. If you run them you will notice that they currently fail when tags are removed. You can see the changes I made to the test file in this PR.

You can run the tests by running the `run_e2e_tests.sh` script in the root project directory. If you need any help getting them going let me know.

Reply:
Ok, I will pull down these tests.  I'll let you know if I have any issues running them.

I agree on the need for unit tests, but was not aware of what was out there so tested manually.  You are correct, I was thinking I only needed to update the tag count on new segments, thinking that the old segment was being deleted anyway.  But I see now that the old segment might still be referenced by other tags, and thus would not be deleted.  I'll work on changing the code to compute the counts for oldTags as well as currentTags.